File: PROCESS SCHEDULING/JS/SRTF.js
class Process {
    constructor(id, at, bt) {
        this.id = id;
        this.at = at;
        this.bt = bt;
        this.st = -1;
        this.ct = 0;
        this.rt = 0;
        this.wt = 0;
        this.tat = 0;
        this.completed = false;
        this.remaining_bt = bt;
    }
}

function findNextProcess(p, currentTime) {
    let minBt = Infinity;
    let minIndex = -1;
    for (let i = 0; i < p.length; i++) {
        if (!p[i].completed && p[i].at <= currentTime) {
            if (p[i].remaining_bt < minBt) {
                minBt = p[i].remaining_bt;
                minIndex = i;
            } else if (p[i].remaining_bt === minBt) {
                if (p[i].at < p[minIndex]?.at) {
                    minIndex = i;
                } else if (p[i].at === p[minIndex]?.at) {
                    if (p[i].id < p[minIndex]?.id) {
                        minIndex = i;
                    }
                }
            }
        }
    }
    return minIndex;
}

function srtfScheduling(inputProcesses) {
    let processes = inputProcesses.map((p, i) => new Process(i + 1, p.at, p.bt));
    let currentTime = 0, totalTat = 0, totalWt = 0, totalIdleTime = 0, completeCount = 0;
    let gChart = [];

    while (completeCount < processes.length) {
        let idx = findNextProcess(processes, currentTime);

        if (idx === -1) {
            currentTime++;
            totalIdleTime++;
        } else {
            let proc = processes[idx];

            if (proc.remaining_bt === proc.bt) {
                proc.st = currentTime;
            }

            gChart.push(proc.id);
            proc.remaining_bt--;
            currentTime++;

            if (proc.remaining_bt === 0) {
                proc.ct = currentTime;
                proc.tat = proc.ct - proc.at;
                proc.wt = proc.tat - proc.bt;
                proc.rt = proc.st - proc.at;

                totalWt += proc.wt;
                totalTat += proc.tat;
                proc.completed = true;
                completeCount++;
            }
        }
    }

    // Output
    console.log("Name: Pranjal Kundu");
    console.log("Section: A1\nRoll no.: 46\n");

    // Gantt Chart
    console.log("\nGantt Chart:");
    let gStr = "", timeStr = "0\t";
    for (let i = 0; i < gChart.length; i++) {
        gStr += `P${gChart[i]}\t|`;
        timeStr += `${i + 1}\t`;
    }
    console.log(gStr);
    console.log(timeStr);

    // Process Table
    console.log("\nPID\tAT\tBT\tCT\tRT\tTAT\tWT");
    for (let p of processes) {
        console.log(`${p.id}\t${p.at}\t${p.bt}\t${p.ct}\t${p.rt}\t${p.tat.toFixed(2)}\t${p.wt.toFixed(2)}`);
    }

    let avgWt = totalWt / processes.length;
    let avgTat = totalTat / processes.length;
    let cpuUtil = ((currentTime - totalIdleTime) / currentTime) * 100;
    let throughput = processes.length / currentTime;

    console.log(`\nCPU Utilization: ${cpuUtil.toFixed(2)}%`);
    console.log(`Throughput: ${throughput.toFixed(2)} processes/unit time`);
    console.log(`Average Waiting Time: ${avgWt.toFixed(2)}`);
    console.log(`Average Turnaround Time: ${avgTat.toFixed(2)}`);
}

// Example usage:
const processes = [
    { at: 0, bt: 5 },
    { at: 1, bt: 3 },
    { at: 2, bt: 1 },
    { at: 4, bt: 2 }
];

srtfScheduling(processes);

File: PROCESS SCHEDULING/CPP/SRTF.cpp
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

struct process {
    int id, at, bt, st, ct, rt, completed, remaining_bt;  
    float wt, tat;
};
typedef struct process process;

// Function to find the next process based on SRTF
int findNextProcess(process p[], int n, int currentTime) {
    int minBt = INT_MAX;  // A large value for comparison
    int minIndex = -1;
    for (int i = 0; i < n; i++) {
        if (p[i].completed == 0 && p[i].at <= currentTime) {
            if (p[i].remaining_bt < minBt) {
                minBt = p[i].remaining_bt;
                minIndex = i;
            } else if (p[i].remaining_bt == minBt) {
                if (p[i].at < p[minIndex].at) {
                    minIndex = i;
                } else if (p[i].at == p[minIndex].at) {
                    if (p[i].id < p[minIndex].id) {
                        minIndex = i;
                    }
                }
            }
        }
    }
    return minIndex;
}

int main() {
    int n;
    printf("Enter the number of processes: ");
    scanf("%d", &n);
    
    process p[n];
    for (int i = 0; i < n; i++) {
        printf("\nEnter arrival time and burst time for process %d: ", i + 1);
        scanf("%d%d", &p[i].at, &p[i].bt);
        p[i].id = i + 1;
        p[i].completed = 0;
        p[i].remaining_bt = p[i].bt; // Initialize remaining burst time
    }

    int gChart[100], gIdx = 0; // Gantt chart array
    int totalTat = 0, totalWt = 0, currentTime = 0, totalIdleTime = 0;
    int complete_count = 0;

    // Simulation loop
    while (complete_count < n) {
        int nextP = findNextProcess(p, n, currentTime);
        
        if (nextP == -1) {
            currentTime++;  // Increment idle time if no process is ready
            totalIdleTime++; // Increment total idle time
        } else {
            // Execute the selected process for 1 time unit
            if (p[nextP].remaining_bt == p[nextP].bt) { // First execution of the process
                p[nextP].st = currentTime; // Start time of the process
            }

            // Store in Gantt chart
            gChart[gIdx++] = p[nextP].id; 
            p[nextP].remaining_bt--; // Decrease remaining burst time
            currentTime++;           // Move current time forward
            
            // If process is completed
            if (p[nextP].remaining_bt == 0) {
                p[nextP].ct = currentTime; // Set completion time
                
                p[nextP].tat = p[nextP].ct - p[nextP].at; // Turnaround time
                p[nextP].wt = p[nextP].tat - p[nextP].bt; // Waiting time
                p[nextP].rt = p[nextP].st - p[nextP].at;  // Response time
                
                totalWt += p[nextP].wt;
                totalTat += p[nextP].tat;
                
                p[nextP].completed = 1;   // Mark as completed
                complete_count++;         // Increment completed count
            }
        }
    }

    // Print the Gantt Chart with times
    printf("\nGantt Chart: \n");
    for (int i = 0; i < gIdx; i++) {
        printf("P%d\t|", gChart[i]);
    }
    printf("\n0\t");

    for (int i = 0; i < gIdx; i++) {
        printf("\t%d", (i == gIdx - 1) ? currentTime : (i + 1));
    }
    printf("\n");

    // Print process details
    printf("\nPID\tAT\tBT\tCT\tRT\tTAT\tWT\n");
    for (int i = 0; i < n; i++) {
        printf("%d\t%d\t%d\t%d\t%d\t%.2f\t%.2f\n", p[i].id, p[i].at, p[i].bt, p[i].ct, p[i].rt, p[i].tat, p[i].wt);
    }

    // Calculate CPU utilization, average waiting time, average turnaround time, and throughput
    float avgWt = (float)totalWt / n;
    float avgTat = (float)totalTat / n;
    float cpu_utilization = (float)(currentTime - totalIdleTime) / currentTime * 100; // Correct CPU utilization
    float throughput = (float)n / currentTime;

    printf("\nCPU Utilization: %.2f%%\n", cpu_utilization);
    printf("Throughput: %.2f processes/unit time\n", throughput);
    printf("Average Waiting Time: %.2f\n", avgWt);
    printf("Average Turnaround Time: %.2f\n", avgTat);

    return 0;
}

